File: firefly-example/src/main/java/com/firefly/example/webscoket/WebSocketServerDemo.java
package com.firefly.example.webscoket;

import com.firefly.$;
import com.firefly.server.http2.router.handler.file.StaticFileHandler;
import com.firefly.utils.concurrent.Scheduler;
import com.firefly.utils.concurrent.Schedulers;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Date;
import java.util.concurrent.TimeUnit;

/**
 * @author Pengtao Qiu
 */
public class WebSocketServerDemo {

    public static void main(String[] args) throws Exception {
        Scheduler scheduler = Schedulers.createScheduler();
        Path path = Paths.get(WebSocketServerDemo.class.getResource("/").toURI());

        $.httpServer()
         .router().get("/static/*").handler(new StaticFileHandler(path.toAbsolutePath().toString()))
         .router().get("/").handler(ctx -> ctx.renderTemplate("template/websocket/index.mustache"))
         .webSocket("/helloWebSocket")
         .onConnect(conn -> {
             Scheduler.Future future = scheduler.scheduleAtFixedRate(() -> conn.sendText("Current time: " + new Date()),
                     0, 1, TimeUnit.SECONDS);
             conn.onClose(c -> future.cancel());
         })
         .onText((text, conn) -> System.out.println("Server received: " + text))
         .listen("localhost", 8080);
    }
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/websocket/WebSocketServerDemo.kt
package com.firefly.example.kotlin.websocket

import com.firefly.codec.http2.model.HttpMethod
import com.firefly.kotlin.ext.http.HttpServer
import com.firefly.server.http2.router.handler.file.StaticFileHandler
import com.firefly.utils.concurrent.Schedulers
import java.nio.file.Paths
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * @author Pengtao Qiu
 */
fun main(args: Array<String>) {
    val scheduler = Schedulers.createScheduler()
    val p = Paths.get(HttpServer::class.java.getResource("/").toURI())

    HttpServer {
        router {
            httpMethod = HttpMethod.GET
            paths = listOf("/favicon.ico", "/static/*")
            handler(StaticFileHandler(p.toAbsolutePath().toString()))
        }

        router {
            httpMethod = HttpMethod.GET
            path = "/"
            asyncHandler { renderTemplate("template/websocket/index.mustache") }
        }

        webSocket("/helloWebSocket") {
            onConnect {
                val future = scheduler.scheduleAtFixedRate(
                        { it.sendText("Current time: " + Date()) },
                        0, 1, TimeUnit.SECONDS)
                it.onClose { future.cancel() }
            }

            onText { text, _ ->
                println("Server received: $text")
            }
        }
    }.listen("localhost", 8080)
}